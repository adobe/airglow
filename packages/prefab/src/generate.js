/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { mapObjIndexed, propOr, identity } from 'ramda';

import togglePrefab from './builtin/toggle.prefab';
import valuePrefab from './builtin/value.prefab';
import listPrefab from './builtin/list.prefab';

const providers = {};

export default mapObjIndexed(
  ({ type, ...options }, name) =>
    propOr(identity, type, providers)({ name, ...options })
);

export const registerProvider = (name, provider) => { providers[name] = provider; };

// register default providers
registerProvider('list', listPrefab);
registerProvider('toggle', togglePrefab);
registerProvider('value', valuePrefab);
