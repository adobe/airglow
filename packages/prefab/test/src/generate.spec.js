/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import generate, { registerProvider } from '../../src/generate';

const doublePrefab = ({ name }) => {
  const selectorInternal = () => {};
  selectorInternal.reduce = s => s;
  return {
    [`${name}A`]: selectorInternal,
    [`${name}B`]: selectorInternal
  };
};

describe('generate prefab', () => {
  it('should process provided configuration', () => {
    expect(generate({
      isOpen: { type: 'toggle' },
      value: { type: 'value', defaultValue: 'test' }
    })).toMatchSnapshot();
  });
  it('handles multiple results from a construct', () => {
    registerProvider('double', doublePrefab);
    expect(generate({
      name: { type: 'double' },
      value: { type: 'double' }
    })).toMatchSnapshot();
  });
});
