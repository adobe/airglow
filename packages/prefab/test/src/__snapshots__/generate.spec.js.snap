// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate prefab handles multiple results from a construct 1`] = `
Object {
  "nameA": [Function],
  "nameB": [Function],
  "valueA": [Function],
  "valueB": [Function],
}
`;

exports[`generate prefab should process provided configuration 1`] = `
Object {
  "isOpen": Object {
    "closeAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_OFF",
    },
    "defaultValue": false,
    "doClose": [Function],
    "doHide": [Function],
    "doOpen": [Function],
    "doSet": [Function],
    "doShow": [Function],
    "doToggle": [Function],
    "doTurnOff": [Function],
    "doTurnOn": [Function],
    "doUnset": [Function],
    "hideAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_OFF",
    },
    "isClosed": [Function],
    "isHidden": [Function],
    "isNotSet": [Function],
    "isOff": [Function],
    "isOn": [Function],
    "isOpen": [Function],
    "isSet": [Function],
    "isVisible": [Function],
    "openAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_ON",
    },
    "reduce": [Function],
    "setAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_ON",
    },
    "showAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_ON",
    },
    "toggleAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_SWITCH",
    },
    "toggleName": "isOpen",
    "turnOffAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_OFF",
    },
    "turnOnAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_ON",
    },
    "unsetAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "isOpen",
      },
      "type": "PREFAB_TOGGLE_OFF",
    },
  },
  "value": Object {
    "changeAction": [Function],
    "defaultValue": "test",
    "doChange": [Function],
    "doReset": [Function],
    "prefabName": "value",
    "reduce": [Function],
    "resetAction": Object {
      "meta": undefined,
      "payload": Object {
        "construct": "value",
      },
      "type": "PREFAB_VALUE_RESET",
    },
    "value": [Function],
  },
}
`;
